development/runtime/testing (#254)
substrate-developer-hub/substrate-developer-hub.github.io: docs/development/module/tests.md

## Unit Testing

Substrate uses the existing [unit testing](https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html) framework provided by Rust. To run tests, the command is 

```bash
cargo test <optional: test_name>
```

## Mock Runtime Environment

> To test a Substrate runtime, construct a mock runtime environment. The [substrate recipes](https://substrate.dev/recipes/testing/mock.html) provides a simple example of using a mocked runtime to test module logic. For more examples, see the `mock.rs` and `test.rs` files in most substrate modules.

Reviewer: ```suggestion
To test a Substrate runtime, construct a mock runtime environment. The [Substrate recipes](https://substrate.dev/recipes/testing/mock.html) provides a simple example of using a mocked runtime to test module logic. For more examples, see the `mock.rs` and `test.rs` files in most substrate pallets.
```

### Mock Runtime Storage

The [`runtime-io`](https://crates.parity.io/sr_io/index.html#enums) crate exposes a [`TestExternalities`](https://crates.parity.io/sr_io/type.TestExternalities.html) implementation frequently used for mocking storage in tests. It is the type alias for an in-memory, hashmap-based externalities implementation in [`substrate_state_machine`](https://crates.parity.io/substrate_state_machine/index.html)] referred to as [`TestExternalities`](https://crates.parity.io/substrate_state_machine/struct.TestExternalities.html).

> In the [basic mock runtime's recipe](https://substrate.dev/recipes/testing/mock.html), an `ExtBuilder` object is defined to build an instance of [`TestExternalities`](https://crates.parity.io/sr_io/type.TestExternalities.html).

Reviewer: ```suggestion
In the [basic mock runtime's recipe](https://substrate.dev/recipes/testing/mock.html), an `ExtBuilder` object is defined to build an instance of [`TestExternalities`](https://substrate.dev/rustdocs/master/sr_io/type.TestExternalities.html).
```

@shawntabrizi asked me the other day to start using the substrate.dev links for rustdocs.

Author: good to know

> ```rust

Reviewer 2: Why is this removed?

Author: added back, must have been a mistake

pub struct ExtBuilder;

impl ExtBuilder {
	pub fn build() -> runtime_io::TestExternalities {
		let mut storage = system::GenesisConfig::default().build_storage::<TestRuntime>().unwrap();
		runtime_io::TestExternalities::from(storage)
	}
}
```

> To create the test environment in unit tests, the build method is called to generate a `TestExternalities` using the default genesis configuration. Then, [`with_externalities`](https://crates.parity.io/substrate_externalities/fn.with_externalities.html) provides the runtime environment in which we may call the module's methods to test that storage, events, and errors behave as expected.

Reviewer: ```suggestion
To create the test environment in unit tests, the build method is called to generate a `TestExternalities` using the default genesis configuration. Then, [`with_externalities`](https://crates.parity.io/substrate_externalities/fn.with_externalities.html) provides the runtime environment in which we may call the pallet's methods to test that storage, events, and errors behave as expected.
```

#### Genesis Config

The previously shown `ExtBuilder::build()` method used the default genesis configuration for building the mock runtime environment. In many cases, it is convenient to set storage before testing.

An example might involve pre-seeding account balances before testing. The `TestRuntime` implements `balances::Trait`, 

```rust
parameter_types! {
	pub const ExistentialDeposit: u64 = 1;
	pub const TransferFee: u64 = 0;
	pub const CreationFee: u64 = 0;
}

impl balances::Trait for TestRuntime {
	type Balance = u64;
	type OnNewAccount = ();
	type OnFreeBalanceZero = ();
	type Event = ();
	type TransferPayment = ();
	type DustRemoval = ();
	type ExistentialDeposit = ExistentialDeposit;
	type TransferFee = TransferFee;
	type CreationFee = CreationFee;

> }

Reviewer 2: I cant make a github suggestion here, but I think we should just keep this minimal:

```
impl balances::Trait for TestRuntime {
	type Balance = u64;
        //--snip//
}
```

All the other parts of the code seem irrelevant for the example.

Author: good call, recently addressed in last commit

```

In the implementation of `system::Trait`, `AccountId` is set to `u64` just like `Balance` shown above. Place `(u64, u64)` pairs in the `balances` vec to seed `(AccountId, Balance)` pairs as the account balances.

```rust
pub fn build(self) -> runtime_io::TestExternalities {
	GenesisConfig {
		balances: Some(balances::GenesisConfig::<TestRuntime>{
			balances: vec![
				(1, 10),
				(2, 20),
				(3, 30),
				(4, 40),
				(5, 50),
				(6, 60)
			],
			vesting: vec![],
		}),
	}.build_storage().unwrap().into()
}
```

Account 1 has balance 10, account 2 has balance 20, and so on.

### Block Production

It will be useful to simulate block production to verify that expected behavior holds during block time dependent changes. 

A simple way of doing this increments the System module's block number between `on_initialize` and `on_finalize` calls from all modules with `System::block_number()` as the sole input.

```rust
fn run_to_block(n: u64) {
	while System::block_number() < n {
		let current_block = System::block_number();
		ExampleModule::on_finalize(current_block);
		System::on_finalize(current_block);
		System::set_block_number(current_block + 1u64.into());

> 		System::on_initialize(current_block);

Reviewer 2: This is wrong. You call `on_initialize` on a old block number...

The syntax I provide does not run into this error.

Author: good catch, 
```suggestion
		System::on_initialize(current_block + 1.into());
```

Reviewer 2: 👎 You should use the old syntax and apply the change to both lines.

Author: To be clear, it is better to call `System::block_number()` four times in this method instead of calling it once and then using that value for each inner method call? 

Can you explain why the suggestion is wrong?

Author: Done

Reviewer: It seems like minimizing storage reads is a good idea how about something like this:

```rust
let mut current_block = System::block_number();
while current_block < n {
  // on finalize calls
  current_block += 1;
  System::set_block_number(current_block);
  // on initialize calls
}
```

		ExampleModule::on_initialize(System::block_number());
	}
}
```

> `on_finalize` and `on_initialize` are only called from `ExampleModule` if the module's trait implements the `sr_primitives::traits::{OnInitialize, OnFinalize}` traits to execute the logic encoded in the runtime methods before and after each block respectively.

Reviewer: Module -> Pallet

Author: addressed in most recent commit

An example might involve pre-seeding account balances before testing.

In the implementation of `system::Trait`, `AccountId` is set to `u64` just like `Balance` shown above. Place `(u64, u64)` pairs in the `balances` vec to seed `(AccountId, Balance)` pairs as the account balances.

```rust, ignore
pub fn build(self) -> runtime_io::TestExternalities {
	GenesisConfig {
		balances: Some(balances::GenesisConfig::<TestRuntime>{
			balances: vec![
				(1, 10),
				(2, 20),
				(3, 30),
				(4, 40),
				(5, 50),
				(6, 60)
			],
			vesting: vec![],
		}),
	}.build_storage().unwrap().into()
}
```

Account 1 has balance 10, account 2 has balance 20, and so on.

### Block Production

It will be useful to simulate block production to verify that expected behavior holds during block time dependent changes. 

A simple way of doing this increments the System module's block number between `on_initialize` and `on_finalize` calls from all modules with `System::block_number()` as the sole input.

```rust, ignore
fn run_to_block(n: u64) {
	while System::block_number() < n {
		ExampleModule::on_finalize(System::block_number());
		System::on_finalize(System::block_number());

> 		System::set_block_number(System::block_number());

Reviewer 2: ```suggestion
		System::set_block_number(System::block_number() + 1);
```

It will be useful to simulate block production to verify that expected behavior holds during block time dependent changes. 

A simple way of doing this increments the System module's block number between `on_initialize` and `on_finalize` calls from all modules with `System::block_number()` as the sole input. While it is important for runtime code to [cache calls](https://substrate.dev/recipes/storage/cache.html) to storage or the system module, the test environment scaffolding should prioritize readability to facilitate future maintenance.

```rust, ignore
fn run_to_block(n: u64) {
	while System::block_number() < n {
		ExampleModule::on_finalize(System::block_number());
		System::on_finalize(System::block_number());
		System::set_block_number(System::block_number() + 1);
		System::on_initialize(System::block_number());
		ExampleModule::on_initialize(System::block_number());
	}
}
```

`on_finalize` and `on_initialize` are only called from `ExampleModule` if the pallet's trait implements the `sr_primitives::traits::{OnInitialize, OnFinalize}` traits to execute the logic encoded in the runtime methods before and after each block respectively.

To use this function in unit tests,

> ```rust, ignore

Reviewer 2: Why are you adding `ignore` to this doc? It only has meaning in RustDocs right?

Author: deleted and IDK, it also has meaning in mdbook

Substrate uses the existing [unit testing](https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html) framework provided by Rust.

To run tests, the command is 

```bash
cargo test <optional: test_name>
```

## Mock Runtime Environment

To support the specific needs of testing a Substrate runtime, you need to construct a "mock" runtime environment. This involves importing the logic of other runtime modules, such as the System module, and establishing runtime storage.

> In general, we suggest looking at and mimicking the various examples of runtime tests in `paint`. This document won't be as comprehensive as what is provided in `paint`.

Reviewer: I think "paint" is outdated. "pallete" is in now.

It will be useful to simulate block production to verify that expected behavior holds during block time dependent changes. 

> A simple way of doing this increments the System module's block number between `on_initialize` and `on_finalize` calls from all modules with `System::block_number()` as the sole input. While it is important for runtime code to [cache calls](https://substrate.dev/recipes/storage/cache.html) to storage or the system module, the test environment scaffolding should prioritize readability to facilitate future maintenance.

Author: The UI is making it unintuitive to reply to this https://github.com/substrate-developer-hub/substrate-developer-hub.github.io/pull/254#discussion_r352620401 so I'll just link it.

@JoshOrndorff It's a test so it doesn't really matter how many calls are made to system or storage. Computation/storage efficiency aren't that important in the test environment scaffolding. With that said, it is definitely not how things should be done for runtime development and I wonder if that might be confusing for readers of the doc. So I added this note.

}
```

> 

Reviewer 2: Please follow the outline of the "next steps" laid out here.

This is one thing I will be firm on keeping consistant (and we have some CSS stuff which checks for these items on pages)

Author: I had renamed it `References and Examples` but ok, I'll change it back

Reviewer 2: Sorry I was not clear enough. I mean the WHOLE format.

* Next Steps
   * Learn More
   * Examples
   * References

If you look at EVERY page it follows this pattern...

Reviewer 2: Notice too the action item are all bullets, so forming your actions as bullets too would be best.

## References and Examples

> The [testing](https://substrate.dev/recipes/base/testing/index.html) chapter of the [Substrate Recipes](https://github.com/substrate-developer-hub/recipes/) compliments the samples shown above, and the [kitchen](https://github.com/substrate-developer-hub/recipes/tree/master/kitchen) provides an environment to run the tests, change the logic, and tinker with the code.

Reviewer 2: Testing link is broken

Author: It won't work until the PR in recipes #87 is merged.
